interoperator/controllers/schedulers/sflabelselectorscheduler/sflabelselectorscheduler_controller.go

```diff
@@ -0,0 +1,251 @@
+limitations under the License.
+*/
+
+package sflabelselectorscheduler
+
+import (
+	"context"
+	"math"
+	"os"
+	"strings"
+
+	osbv1alpha1 "github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/api/osb/v1alpha1"
+	resourcev1alpha1 "github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/api/resource/v1alpha1"
+
+	"github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/internal/config"
+	rendererFactory "github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/internal/renderer/factory"
+	"github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/pkg/cluster/registry"
+	"github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/pkg/constants"
+	"github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/pkg/errors"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SFLabelSelectorScheduler reconciles a SFLabelSelectorScheduler object
+type SFLabelSelectorScheduler struct {
+	client.Client
+	Log             logr.Logger
+	scheme          *runtime.Scheme
+	clusterRegistry registry.ClusterRegistry
+}
+
+// Reconcile schedules the SFServiceInstance to one SFCluster and sets the ClusterID in
+// SFServiceInstance.Spec.ClusterID. It chooses the destination cluster based on clusterSelector
+// template provided in the plan.
+func (r *SFLabelSelectorScheduler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("sfserviceinstance", req.NamespacedName)
+
+	instance := &osbv1alpha1.SFServiceInstance{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if instance.Spec.ClusterID == "" {
+		labelSelector, err := getLabelSelectorString(instance, *r)
+		if err != nil {
+			log.Info("Failed to get labelSelector string..", "error", err, "labelSelector", labelSelector)
+			return ctrl.Result{}, err
+		}
+
+		clusterID, err := r.schedule(instance, labelSelector)
+		if err != nil {
+			log.Error(err, "Failed to schedule ", "labelSelector", labelSelector, "clusterID", clusterID)
+			if errors.SchedulerFailed(err) {
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, err
+		}
+
+		if clusterID != "" {
+			log.Info("Setting clusterID", "clusterID", clusterID)
+			instance.Spec.ClusterID = clusterID
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				return r.Update(ctx, instance)
```

Comment:
On conflict we have fetch again and update

Reply:
Done.